### PR TITLE
Move Raid Luks and Integrity to context manager

### PR DESCRIPTION
**Pull request by schaefi**
Change the RaidDevice, LuksDevice and IntegrityDevice classes to context manager. All code using this classes moved to:
```python
with RaidDevice(...) as raid:
    raid.some_member()

with LuksDevice(...) as luks:
    luks.some_member()

with IntegrityDevice(...) as integrity:
    integrity.some_member()
```

In addition to the context manager move, the use of them has been refactored into a nested structure of supported device layouts inside the disk builder.

This is related to Issue #2412

### CODE REVIEW

File: kiwi/builder/disk.py [hunk 1]
```diff
@@ -349,26 +349,66 @@ def create_disk(self) -> Result:
...
+                            self._process_build(device_map, disk)
+                elif self.luks is not None:
+                    with self._luks_instance(device_map) as luks_root:
+                        # create luks on root dev
+                        device_map = self._map_luks(device_map, luks_root)
+                        # build bootable luks disk
+                        self._process_build(device_map, disk)
+                else:
+                    # build bootable disk
+                    self._process_build(device_map, disk)
```
Comment (dcermak) [inline kiwi/builder/disk.py:411]: This part looks rather convoluted and a bit too deeply nested for my taste. Also, the `self._process_build()` call at the end of every exit from the context manager looks like there should be a better way to do this.

Have you considered using a single exit stack, use `push()` to add all the Raid/Integrity/LUKS but thereby preserve the previous & much easier to follow code structure?

**Review by dcermak — CHANGES_REQUESTED** [review]
I think the new code in disk.py could use a bit of simplification

File: kiwi/builder/disk.py (hunk 1 above)
Comment (schaefi) [reply kiwi/builder/disk.py:411]: I knew you would say that ;) I also thought while writing it that this is not optimal. However, from the conversation about context managers I had the impression that you wanted this type of nested structure, no ?

I'm fine with an ExitStack that is much better. Will adapt the PR

Comment (schaefi) [reply kiwi/builder/disk.py]: done

Comment (dcermak) [reply kiwi/builder/disk.py]: > However, from the conversation about context managers I had the impression that you wanted this type of nested structure, no ?

No, I actually don't prefer this deep nesting that context managers cause. Hence I am very happy that the exit stack exists, as it allows us to combine the advantages of context managers and not having a five level deep nesting.

File: kiwi/builder/disk.py [hunk 2]
```diff
@@ -703,6 +710,9 @@ def _map_root_filesystem(
         self.storage_map['system'] = filesystem
 
     def _process_build(self, device_map: Dict, disk: Disk) -> None:
+        # create system layout for root system
+        self._create_system_instance(device_map)
```
Comment (dcermak) [inline kiwi/builder/disk.py:714]: Why is this suddenly necessary?

**Review by dcermak — APPROVED** [review]
Looks much better, thanks!

Just added one question about a code addition

File: kiwi/builder/disk.py (hunk 2 above)
Comment (schaefi) [reply kiwi/builder/disk.py:714]: It was there before. The call moved into the function as it imho makes more sense to have it there

Comment (schaefi) [reply kiwi/builder/disk.py:714]: see the red lines from the diff, you will see the call as an extra line but it only makes sense in combination with the build of the disk

Comment (dcermak) [reply kiwi/builder/disk.py:714]: Ah, sorry, I completely missed it :facepalm:

**Pull request merged**